diff --git a/fhi_lib/lib/api/xran_fh_o_ru.h b/fhi_lib/lib/api/xran_fh_o_ru.h
index 7812399..8bdb539 100644
--- a/fhi_lib/lib/api/xran_fh_o_ru.h
+++ b/fhi_lib/lib/api/xran_fh_o_ru.h
@@ -28,7 +28,7 @@
  * @author Intel Corporation
  *
  **/
-
+#pragma once
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -48,6 +48,27 @@ extern "C" {
 #include <getopt.h>
 #include <unistd.h>
 #include "xran_fh_o_du.h"
+#include "xran_transport.h"
+
+typedef int (*process_uplane_fn)(struct rte_mbuf *pkt, void *handle,
+                                 struct xran_eaxc_info *p_cid,
+                                 uint16_t xport_id,
+                                 struct xran_sense_of_time *sense_of_time);
+typedef int32_t (*process_cplane_fn)(struct rte_mbuf *pkt, void *handle, uint16_t xport_id,
+                                     struct xran_sense_of_time *sense_of_time);
+typedef void (*sym_ota_fn)(void * handle, struct xran_sense_of_time *sense_of_time);
+
+void xran_hook_install(void *pHandle, process_uplane_fn process_uplane_fn_p,
+                       void *process_uplane_fn_args,
+                       process_cplane_fn process_cplane_fn_p,
+                       void *process_cplane_fn_args,
+                       sym_ota_fn sym_ota_fn_p,
+                       void *sym_ota_fn_args,
+                       int mu);
+void xran_hook_schedule_packet(void *pHandle, struct rte_mbuf *mbuf, int port,
+                               enum xran_pkt_dir direction, int ru_port_id,
+                               int slot, int symbol);
+int xran_hook_send_packet(void *pHandle, struct rte_mbuf *mbuf, int port, enum xran_pkt_dir direction, int ru_port_id);
 
 /**
  * @ingroup
@@ -78,4 +99,4 @@ int32_t xran_5g_bfw_config(void * pHandle, struct xran_buffer_list *pSrcRxCpBuff
 
 #ifdef __cplusplus
 }
-#endif
\ No newline at end of file
+#endif
diff --git a/fhi_lib/lib/api/xran_pkt.h b/fhi_lib/lib/api/xran_pkt.h
index 314b8d6..478f092 100644
--- a/fhi_lib/lib/api/xran_pkt.h
+++ b/fhi_lib/lib/api/xran_pkt.h
@@ -339,6 +339,11 @@ struct xran_pkt_comm_hdr
     struct xran_ecpri_hdr ecpri_hdr; /**< eCPRI Transport Header */
 } __rte_packed;
 
+enum xran_mbuf_mem_op_id {
+    MBUF_KEEP,
+    MBUF_FREE
+};
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/fhi_lib/lib/ethernet/ethdi.h b/fhi_lib/lib/ethernet/ethdi.h
index 5a5bf80..e32620c 100644
--- a/fhi_lib/lib/ethernet/ethdi.h
+++ b/fhi_lib/lib/ethernet/ethdi.h
@@ -117,11 +117,6 @@ struct xran_ethdi_ctx {
     uint64_t rx_vf_queue_cnt[XRAN_VF_MAX][XRAN_VF_QUEUE_MAX];
 };
 
-enum xran_mbuf_mem_op_id {
-    MBUF_KEEP,
-    MBUF_FREE
-};
-
 extern enum xran_if_state xran_if_current_state;
 
 static inline struct xran_ethdi_ctx *xran_ethdi_get_ctx(void)
diff --git a/fhi_lib/lib/src/xran_dev.c b/fhi_lib/lib/src/xran_dev.c
index c939edc..9037091 100644
--- a/fhi_lib/lib/src/xran_dev.c
+++ b/fhi_lib/lib/src/xran_dev.c
@@ -502,3 +502,118 @@ xran_init_vfs_mapping(void *pHandle)
 
     return (XRAN_STATUS_SUCCESS);
 }
+
+#define MAX_PACKETS_SENT_PER_SYMBOL 16 // has to be power of 2
+
+// Process the scheduled packets, done once per symbol at OTA time
+void xran_hook_process_tx_packets(void* pHandle, int port, int slot, int symbol) {
+  struct xran_device_ctx *p_dev_ctx = (struct xran_device_ctx *)pHandle;
+  int ring_index = slot % XRAN_N_FE_BUF_LEN;
+  struct rte_ring *ring = p_dev_ctx->hook_cfg.tx_rings[port][ring_index][symbol];
+  void* mbufs[MAX_PACKETS_SENT_PER_SYMBOL];
+  int dequeued = rte_ring_dequeue_burst(ring, mbufs, MAX_PACKETS_SENT_PER_SYMBOL, NULL);
+  struct xran_ethdi_ctx *ctx = xran_ethdi_get_ctx();
+  struct rte_ring *tx_ring = ctx->tx_ring[port];
+  int enqueued = rte_ring_enqueue_burst(tx_ring, mbufs, dequeued, NULL);
+  if (enqueued != dequeued) {
+    rte_panic("Expected to be able to enqueue all the packets on the TX ring\n");
+  }
+}
+
+typedef struct {
+  void* pHandle;
+  int port;
+  sym_ota_fn sym_ota_fn;
+  void *sym_ota_fn_args;
+} callback_args_t;
+
+int32_t send_packets(void* arg, struct xran_sense_of_time* p_sense_of_time) {
+  callback_args_t *cb_args = arg;
+  int mu = xran_get_conf_numerology(cb_args->pHandle);
+  int slot_in_frame = p_sense_of_time->nSlotIdx + p_sense_of_time->nSubframeIdx * (1 << mu);
+  xran_hook_process_tx_packets(cb_args->pHandle, cb_args->port, slot_in_frame, p_sense_of_time->nSymIdx);
+  if (cb_args->sym_ota_fn) {
+    cb_args->sym_ota_fn(cb_args->sym_ota_fn_args, p_sense_of_time);
+  }
+}
+
+void xran_hook_install(
+  void *pHandle,
+  process_uplane_fn process_uplane_fn_p,
+  void *process_uplane_fn_args,
+  process_cplane_fn process_cplane_fn_p,
+  void *process_cplane_fn_args,
+  sym_ota_fn sym_ota_fn_p,
+  void *sym_ota_fn_args,
+  int mu)
+{
+  static bool installed = false;
+  if (installed) rte_panic("Can only install once for one numerology\n");
+  installed = true;
+
+  // Workaround - indicate XRAN memory is ready so packet processor starts
+  struct xran_device_ctx *p_dev_ctx = (struct xran_device_ctx *)pHandle;
+  p_dev_ctx->xran2phy_mem_ready = 1;
+
+  if (mu != xran_get_conf_numerology(pHandle)) {
+    rte_panic("Can only use numerlogy %d as configured during init\n", xran_get_conf_numerology(pHandle));
+  }
+
+  p_dev_ctx->hook_cfg.process_uplane_fn = process_uplane_fn_p;
+  p_dev_ctx->hook_cfg.process_uplane_fn_args = process_uplane_fn_args;
+  p_dev_ctx->hook_cfg.process_cplane_fn = process_cplane_fn_p;
+  p_dev_ctx->hook_cfg.process_cplane_fn_args = process_cplane_fn_args;
+
+  struct xran_ethdi_ctx *ctx = xran_ethdi_get_ctx();
+  for (int port = 0; port < ctx->io_cfg.num_vfs; port++) {
+    int ring_index = 0;
+    for (int slot = 0; slot < XRAN_N_FE_BUF_LEN; slot++) {
+      for (int symbol = 0; symbol < XRAN_SYMBOLNUMBER_MAX; symbol++) {
+        char ring_name[RTE_RING_NAMESIZE];
+        snprintf(ring_name, RTE_DIM(ring_name), "%s_%d_%d", "hook_tx_ring", port, ring_index++);
+        struct rte_ring *ring = rte_ring_create(ring_name, MAX_PACKETS_SENT_PER_SYMBOL, SOCKET_ID_ANY, RING_F_SC_DEQ);
+        if (ring == NULL) rte_panic("Cannot allocate rte_ring\n");
+        p_dev_ctx->hook_cfg.tx_rings[port][slot][symbol] = ring;
+      }
+    }
+  }
+
+  // Setup a callback each OTA symbol to send packets from per-symbol rings into the TX ring
+  static struct xran_sense_of_time xran_sense_of_time[XRAN_SYMBOLNUMBER_MAX];
+  static callback_args_t cb_args = {0};
+  cb_args.port = 0;
+  cb_args.pHandle = pHandle;
+  cb_args.sym_ota_fn = sym_ota_fn_p;
+  cb_args.sym_ota_fn_args = sym_ota_fn_args;
+  for (int i = 0; i < XRAN_SYMBOLNUMBER_MAX; i++) {
+    xran_reg_sym_cb(pHandle, send_packets, &cb_args, &xran_sense_of_time[i], i, XRAN_CB_SYM_OTA_TIME);
+  }
+}
+
+// Schedule packet for slot and symbol
+void xran_hook_schedule_packet(void *pHandle, struct rte_mbuf *mbuf, int port,
+                               enum xran_pkt_dir direction, int ru_port_id,
+                               int slot, int symbol) {
+  void *ret = rte_pktmbuf_prepend(mbuf, sizeof(struct rte_ether_hdr));
+  if (ret == NULL) rte_panic("not enough headroom for ethernet header");
+  struct xran_device_ctx *p_dev_ctx = (struct xran_device_ctx *)pHandle;
+  int vf_id = xran_map_ecpriPcid_to_vf(pHandle, direction, 0, ru_port_id);
+  struct xran_ethdi_ctx *ctx = xran_ethdi_get_ctx();
+
+  mbuf->port = ctx->io_cfg.port[vf_id];
+  xran_add_eth_hdr_vlan(&ctx->entities[vf_id][ID_O_RU], ETHER_TYPE_ECPRI, mbuf);
+
+  struct rte_ring *ring = p_dev_ctx->hook_cfg.tx_rings[port][slot][symbol];
+  int res = rte_ring_enqueue(ring, mbuf);
+  if (res != 0) rte_panic("Cannot enqueue to ring");
+}
+
+int xran_hook_send_packet(void *pHandle, struct rte_mbuf *mbuf, int port, enum xran_pkt_dir direction, int ru_port_id) {
+  void *ret = rte_pktmbuf_prepend(mbuf, sizeof(struct rte_ether_hdr));
+  if (ret == NULL) rte_panic("not enough headroom for ethernet header");
+  int vf_id = xran_map_ecpriPcid_to_vf(pHandle, direction, 0, ru_port_id);
+  if (xran_ethdi_mbuf_send(mbuf, ETHER_TYPE_ECPRI, vf_id) == 1) {
+    return 0;
+  }
+  return 1;
+}
diff --git a/fhi_lib/lib/src/xran_dev.h b/fhi_lib/lib/src/xran_dev.h
index 778b38d..9a19ab5 100644
--- a/fhi_lib/lib/src/xran_dev.h
+++ b/fhi_lib/lib/src/xran_dev.h
@@ -40,6 +40,7 @@ extern "C" {
 #include <rte_timer.h>
 
 #include "xran_fh_o_du.h"
+#include "xran_fh_o_ru.h"
 #include "xran_prach_cfg.h"
 #include "xran_up_api.h"
 #include "xran_cp_api.h"
@@ -186,6 +187,17 @@ struct xran_prb_elm_proc_info_t {
     uint8_t   numSymsRemaining; /**< Number of symbols for DL CP transmission remaining in this slot */
 };
 
+typedef struct {
+  // Used to replace XRAN packet processing functions
+  process_uplane_fn process_uplane_fn;
+  void *process_uplane_fn_args;
+  process_cplane_fn process_cplane_fn;
+  void *process_cplane_fn_args;
+
+  // Used to schedule TX in the future.
+  struct rte_ring *tx_rings[XRAN_VF_MAX][XRAN_N_FE_BUF_LEN][XRAN_SYMBOLNUMBER_MAX];
+} hook_cfg_t;
+
 struct __rte_cache_aligned xran_device_ctx
 {
     uint8_t sector_id;
@@ -326,6 +338,7 @@ struct __rte_cache_aligned xran_device_ctx
     uint8_t technology[XRAN_MAX_DSS_PERIODICITY];   /**< technology array represents slot is LTE(0)/NR(1) */
     /* Keeps track of how many sections are processed while parsing C-plan packet */
     uint8_t sectiondb_elm[XRAN_MAX_SECTIONDB_CTX][XRAN_DIR_MAX][XRAN_COMPONENT_CARRIERS_MAX][XRAN_MAX_ANTENNA_NR * 2 + XRAN_MAX_ANT_ARRAY_ELM_NR];
+    hook_cfg_t hook_cfg;
 };
 
 struct xran_eaxcid_config *xran_get_conf_eAxC(void *pHandle);
diff --git a/fhi_lib/lib/src/xran_main.c b/fhi_lib/lib/src/xran_main.c
index db03d94..1dfb5f7 100644
--- a/fhi_lib/lib/src/xran_main.c
+++ b/fhi_lib/lib/src/xran_main.c
@@ -1680,9 +1680,31 @@ int32_t handle_ecpri_ethertype(struct rte_mbuf* pkt_q[], uint16_t xport_id, stru
         }
     else
 {
+    struct xran_sense_of_time sense_of_time;
+    if (p_dev_ctx->hook_cfg.process_uplane_fn || p_dev_ctx->hook_cfg.process_cplane_fn) {
+        int mu = p_dev_ctx->fh_cfg.frame_conf.nNumerology;
+        int32_t slots = xran_lib_ota_sym_idx[p_dev_ctx->xran_port_id] / XRAN_NUM_OF_SYMBOL_PER_SLOT;
+        int num_slots_per_frame = 10 << mu;
+        int num_slots_per_subframe = 1 << mu;
+        int frame = slots / num_slots_per_frame;
+        int subframe = (slots % num_slots_per_frame) / num_slots_per_subframe;
+        int slot = slots - frame * num_slots_per_frame - subframe * num_slots_per_subframe;
+        int symbol = xran_lib_ota_sym_idx[p_dev_ctx->xran_port_id] % XRAN_NUM_OF_SYMBOL_PER_SLOT;
+        sense_of_time.tti_counter = xran_lib_ota_sym_idx[p_dev_ctx->xran_port_id];
+        sense_of_time.nFrameIdx = xran_getSfnSecStart() + frame;
+        sense_of_time.nSymIdx = symbol;
+        sense_of_time.nSubframeIdx = subframe;
+        sense_of_time.nSlotIdx = slot;
+        sense_of_time.type_of_event = XRAN_CB_SYM_OTA_TIME;
+        sense_of_time.nSecond = timing_get_current_second();
+    }
         for (i = 0; i < num_data; i++)
     {
-            ret = process_mbuf(pkt_data[i], (void*)p_dev_ctx, p_cid);
+            if (p_dev_ctx->hook_cfg.process_uplane_fn) {
+                 ret = p_dev_ctx->hook_cfg.process_uplane_fn(pkt_data[i], p_dev_ctx->hook_cfg.process_uplane_fn_args, p_cid, xport_id, &sense_of_time);
+            } else {
+                ret = process_mbuf(pkt_data[i], (void*)p_dev_ctx, p_cid);
+            }
             if (ret == MBUF_FREE)
                 rte_pktmbuf_free(pkt_data[i]);
     }
@@ -1692,7 +1714,11 @@ int32_t handle_ecpri_ethertype(struct rte_mbuf* pkt_q[], uint16_t xport_id, stru
             t1 = MLogXRANTick();
             if (p_dev_ctx->fh_init.io_cfg.id == O_RU)
         {
-                ret = process_cplane(pkt_control[i], (void*)p_dev_ctx);
+                if (p_dev_ctx->hook_cfg.process_cplane_fn) {
+                  ret = p_dev_ctx->hook_cfg.process_cplane_fn(pkt_control[i], p_dev_ctx->hook_cfg.process_cplane_fn_args, xport_id, &sense_of_time);
+                } else {
+                  ret = process_cplane(pkt_control[i], (void*)p_dev_ctx);
+                }
                 p_dev_ctx->fh_counters.rx_counter++;
                 if (ret == MBUF_FREE)
                     rte_pktmbuf_free(pkt_control[i]);
@@ -3987,10 +4013,12 @@ xran_start(void *pHandle)
     char buff[100];
     int i;
     struct xran_device_ctx * p_xran_dev_ctx = xran_dev_get_ctx();
-    struct xran_prb_map * prbMap0 = (struct xran_prb_map *) p_xran_dev_ctx->sFrontHaulRxPrbMapBbuIoBufCtrl[0][0][0].sBufferList.pBuffers->pData;
-    for(i = 0; i < XRAN_MAX_SECTIONS_PER_SLOT && i < prbMap0->nPrbElm; i++)
-    {
-        p_xran_dev_ctx->numSetBFWs_arr[i] = prbMap0->prbMap[i].bf_weight.numSetBFWs;
+    if (p_xran_dev_ctx->sFrontHaulRxPrbMapBbuIoBufCtrl[0][0][0].sBufferList.pBuffers) {
+        struct xran_prb_map * prbMap0 = (struct xran_prb_map *) p_xran_dev_ctx->sFrontHaulRxPrbMapBbuIoBufCtrl[0][0][0].sBufferList.pBuffers->pData;
+        for(i = 0; i < XRAN_MAX_SECTIONS_PER_SLOT && i < prbMap0->nPrbElm; i++)
+        {
+            p_xran_dev_ctx->numSetBFWs_arr[i] = prbMap0->prbMap[i].bf_weight.numSetBFWs;
+        }
     }
 
     if(xran_get_if_state() == XRAN_RUNNING) {
